### ISSUE TITLE
Keepalive handler not triggered on state change?

### ISSUE THREAD

**Issue by xyntrix** [sensu/sensu#1097]
We are seeing issues where upon state change, the keepalive handler is not triggering a handler call for change from critical to clear, and falls back into waiting for n occurrences go through.

Also, it looks like the keepalive handler might have a default occurrence of 1 hour?

```
{"timestamp":"2015-11-11T14:12:33.034696-0700","level":"info","message":"handler output","handler":{"type":"pipe","command":"/etc/
sensu/conf.d/enterprise/handlers/notification/rest-alarm.rb","name":"enterprise_restalarm"},"output":["only handling every 60 occurrences: somehost005/keepalive\n"]}
{"timestamp":"2015-11-11T14:14:06.781008-0700","level":"info","message":"handler output","handler":{"type":"pipe","command":"/etc/
sensu/conf.d/enterprise/handlers/notification/rest-alarm.rb","name":"enterprise_restalarm"},"output":["only handling every 60 occurrences: somehost005/keepalive\n"]}
{"timestamp":"2015-11-11T14:16:42.316069-0700","level":"info","message":"handler output","handler":{"type":"pipe","command":"/etc/sensu/conf.d/enterprise/handlers/notification/rest-alarm.rb","name":"enterprise_restalarm"},"output":["endpoint_url: http://restalarms.example.tld/restalarm/\n","message_body: {:ip=>\"1.2.3.4\", :hostname=>\"somehost005\", :service=>\"keepalive\", :status=>\"critical\", :comment=>\"No keepalive sent from client for 2074 seconds (>=300)\", :heartbeat=>0, :impactedip=>\"1.2.3.4\", :maintenance=>false, :documentid=>0}\n"]}
{"timestamp":"2015-11-11T14:19:17.941590-0700","level":"info","message":"handler output","handler":{"type":"pipe","command":"/etc/
sensu/conf.d/enterprise/handlers/notification/rest-alarm.rb","name":"enterprise_restalarm"},"output":["only handling every 60 occurrences: somehost005/keepalive\n"]}
{"timestamp":"2015-11-11T14:20:52.373793-0700","level":"info","message":"handler output","handler":{"type":"pipe","command":"/etc/
sensu/conf.d/enterprise/handlers/notification/rest-alarm.rb","name":"enterprise_restalarm"},"output":["only handling every 60 occurrences: somehost005/keepalive\n"]}
{"timestamp":"2015-11-11T14:23:23.991554-0700","level":"info","message":"handler output","handler":{"type":"pipe","command":"/etc/
sensu/conf.d/enterprise/handlers/notification/rest-alarm.rb","name":"enterprise_restalarm"},"output":["only handling every 60 occurrences: somehost005/keepalive\n"]}
{"timestamp":"2015-11-11T14:25:46.418762-0700","level":"info","message":"handler output","handler":{"type":"pipe","command":"/etc/
sensu/conf.d/enterprise/handlers/notification/rest-alarm.rb","name":"enterprise_restalarm"},"output":["only handling every 60 occurrences: somehost005/keepalive\n"]}
{"timestamp":"2015-11-11T14:27:27.016141-0700","level":"info","message":"handler output","handler":{"type":"pipe","command":"/etc/
sensu/conf.d/enterprise/handlers/notification/rest-alarm.rb","name":"enterprise_restalarm"},"output":["only handling every 60 occurrences: somehost005/keepalive\n"]}
```

**Comment by portertech** [comment]
@xyntrix up to the event handler, w/ the sensu-plugin logic you could just set occurrences and refresh in your client definitions, e.g. `"client": {"keepalive": {"occurrences": 2, "refresh": 1200}}`.

**Comment by xyntrix** [comment]
Hmm.  Shouldn't this, by default, invoke the handler upon state change?   I'm expecting that if I moved from an alarm state (2)  to a clear state (0), the handler will immediately invoke -- but what I see happening is it not immediately invoking the handler.

**Comment by portertech** [comment]
@xyntrix sounds like the handler itself is filtering the event and not acting on it.

**Comment by xyntrix** [comment]
Yeah, this is a weird one.  Testing individually it works as expected:  we ship with the handler on state change.  We've noticed this weird behavior when we have a network outage/degradation:  the keepalive queue will be at a very high level (hundreds of thousands / millions of messages).   If we purge that queue, is when we stometimes see this behavior.  We expect that even if we throw away the keepalive queue and start over, the handler respects last-known vs current state and handles state change accordingly.  Maybe there is an inconsistent state getting stored in redis.. we'll keep digging.

**Comment by xyntrix** [comment]
This was for a single server.. Here's what I'm seeing from the debug logs:   https://gist.github.com/224a66b20da2629db172


fields are:
- timestamp from sensu-server log
- type of action
- issued timestamp
- output message
- status

We flushed the results and keepalive rabbitmq queues just prior to restarting our sensu servers around :6:57 we restarted out sensu servers.

I'm guessing we dumped the clear messages on the floor, but I want to understand how this works.

Hypothesis: is the redis state updated with the state of the client before we publish a "call handler" message?  If so, then if the call-a-handler message is lost -- maybe is why we never actually invoke the clear on the handler?
